streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamSelfJoin.java

```diff
@@ -0,0 +1,144 @@
+            if (StreamStreamJoinUtil.skipRecord(record, LOG, droppedRecordsSensor, context())) {
+                return;
+            }
+
+            final long inputRecordTimestamp = record.timestamp();
+            long timeFrom = Math.max(0L, inputRecordTimestamp - joinThisBeforeMs);
+            long timeTo = Math.max(0L, inputRecordTimestamp + joinThisAfterMs);
+            boolean emittedJoinWithSelf = false;
+            final Record selfRecord = record
+                .withValue(joinerThis.apply(record.key(), record.value(), (V2) record.value()))
+                .withTimestamp(inputRecordTimestamp);
+            sharedTimeTracker.advanceStreamTime(inputRecordTimestamp);
+
+            // Join current record with other
+            try (final WindowStoreIterator<V2> iter = windowStore.fetch(
+                record.key(), timeFrom, timeTo)) {
+                while (iter.hasNext()) {
+                    final KeyValue<Long, V2> otherRecord = iter.next();
+                    final long otherRecordTimestamp = otherRecord.key;
+
+                    // Join this with other
+                    context().forward(
+                        record.withValue(joinerThis.apply(
+                                record.key(), record.value(), otherRecord.value))
+                            .withTimestamp(Math.max(inputRecordTimestamp, otherRecordTimestamp)));
+                }
+            }
+
+            // Needs to be in a different loop to ensure correct ordering of records where
+            // correct ordering means it matches the output of an inner join.
+            timeFrom = Math.max(0L, inputRecordTimestamp - joinOtherBeforeMs);
+            timeTo = Math.max(0L, inputRecordTimestamp + joinOtherAfterMs);
+            try (final WindowStoreIterator<V2> iter2 = windowStore.fetch(
+                record.key(), timeFrom, timeTo)) {
+                while (iter2.hasNext()) {
+                    final KeyValue<Long, V2> otherRecord = iter2.next();
+                    final long otherRecordTimestamp = otherRecord.key;
+                    final long maxRecordTimestamp = Math.max(inputRecordTimestamp, otherRecordTimestamp);
+
+                    // This is needed so that output records follow timestamp order
+                    // Join this with self
+                    if (inputRecordTimestamp < maxRecordTimestamp && !emittedJoinWithSelf) {
+                        emittedJoinWithSelf = true;
+                        context().forward(selfRecord);
+                    }
+
+                    // Join other with current record
+                    context().forward(
+                        record
+                            .withValue(joinerThis.apply(record.key(), (V1) otherRecord.value, (V2) record.value()))
+                            .withTimestamp(Math.max(inputRecordTimestamp, otherRecordTimestamp)));
+                }
+            }
+
+            // Join this with self
```

Comment:
@guozhangwang I tried doing the change in `KStreamJoinWindow` (inverting the lines)
```
 if (record.key() != null) {
                context().forward(record);
                // Every record basically starts a new window. We're using a window store mostly for the retention.
                window.put(record.key(), record.value(), record.timestamp());
            }
```

but it results in the self-record to get emitted multiple times*. This means that we still need logic to check if it was already emitted to emit only once. Which doesn't make the code any clearer so I decided against it. 

* This happens because the record is already in the state store, so it joins with itself in the first loop but also in the second.

Reply:
Got it, I missed that part.. thanks for checking and letting me know @vpapavas